File: 프로그래머스/2/17679. ［1차］ 프렌즈4블록/［1차］ 프렌즈4블록.py
def solution(m, n, board):
    answer = 0
    board = [list(row) for row in board]
    
    while True:
        remove = set()

        for i in range(m - 1):
            for j in range(n - 1):
                if board[i][j] and board[i][j] == board[i][j+1] and board[i][j] == board[i+1][j] and board[i][j] == board[i+1][j+1]:
                    remove.update([(i, j), (i, j + 1), (i + 1, j), (i + 1, j + 1)])

        if not remove:
            break

        answer += len(remove)
        for a, b in remove:
            board[a][b] = ''
            
        for col in range(n):
            stack = [board[row][col] for row in range(m) if board[row][col] != '']
            # 빈 공간을 채우고 블록을 아래로 이동
            for row in range(m - len(stack)):
                board[row][col] = ''
            for row in range(m - len(stack), m):
                board[row][col] = stack[row - (m - len(stack))]
        
    return answer

File: 프로그래머스/2/17679. ［1차］ 프렌즈4블록/［1차］ 프렌즈4블록.java
class Solution {
    public int solution(int m, int n, String[] board) {
        char[][] map = new char[m][n];
        for (int i = 0; i < m; i++)
            map[i] = board[i].toCharArray();
        
        int remove = 0;
        while (true) {
            boolean[][] marked = new boolean[m][n];
            boolean match = false;
            
            for (int i = 0; i < m-1; i++) {
                for (int j = 0; j < n-1; j++) {
                    char current = map[i][j];
                    if (current == ' ') 
                        continue;
                    if (
                        map[i][j+1] == current &&
                        map[i+1][j] == current &&
                        map[i+1][j+1] == current
                    ) {
                        marked[i][j] = true;
                        marked[i+1][j] = true;
                        marked[i][j+1] = true;
                        marked[i+1][j+1] = true;
                        match = true;
                    }
                }
            }
            if (!match)
                break;
            for (int i = 0; i < m; i++) {
                for (int j = 0; j < n; j++) {
                    if (marked[i][j]) {
                        map[i][j] = ' ';
                        remove++;
                    }
                }
            }
            for (int j = 0; j < n; j++) {
                for (int i = m-1; i >= 0; i--) {
                    if (map[i][j] == ' ') {
                        int k = i - 1;
                        while (k >= 0 && map[k][j] == ' ') {
                            k--;
                        }
                        if (k >= 0) {
                            map[i][j] = map[k][j];
                            map[k][j] = ' ';
                        }
                    }
                }
            }
        }
        
        return remove;
    }
}